File: Sample/Source/Windows/WindowShop.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
from typing import Callable, Dict, List, Optional, Tuple, Any
from Engine import (
    Color,
    Input,
    Pair,
    Text,
    Texture,
    UI,
    Vector2f,
)
from Engine.UI import Canvas, ListView
from Engine.UI.Base import FunctionalBase
from Engine.UI.FunctionalUI import FImage, FPlainText
from Global import Manager, System as GlobalSystem
from .Base import WindowSelectable
from .WindowCommand import WindowCommand
from .. import Data
from ..System import System as GameSystem


SHOP_MODE_BUY = "buy"
SHOP_MODE_SELL = "sell"

_SHOP_COMMAND_HEIGHT = 64
_SHOP_ITEM_SIZE = 352
_SHOP_WIDTH = 352
_SHOP_ITEM_ROW_HEIGHT = 32
_SHOP_ITEM_COLUMNS = 2
_SHOP_DISABLED_ALPHA = 120
_SHOP_DISABLED_TEXT_COLOUR = Color(160, 160, 160, 255)
_SHOP_TEXT_SIZE = 12
_SHOP_TEXT_RIGHT_PAD = 2
_SHOP_TEXT_Y_OFFSET = 17

_DEFAULT_COMMAND_RECT: Tuple[Pair[int], Pair[int]] = ((144, 32), (_SHOP_WIDTH, _SHOP_COMMAND_HEIGHT))
_DEFAULT_ITEM_RECT: Tuple[Pair[int], Pair[int]] = ((144, 96), (_SHOP_ITEM_SIZE, _SHOP_ITEM_SIZE))


class _ShopCell(Canvas, FunctionalBase):
    r"""\brief Shop item cell displaying an icon and right-aligned numeric text."""

    def __init__(self, width: int, iconTexture: Optional[Texture], textValue: str, available: bool) -> None:
        r"""\brief Construct a shop item cell.

        - \param width Cell width in logical UI units.
        - \param iconTexture Item icon texture, or None.
        - \param textValue Right-aligned price/count text.
        - \param available Whether the item can currently be confirmed.
        """
        Canvas.__init__(self, ((0, 0), (width, _SHOP_ITEM_ROW_HEIGHT)))
        FunctionalBase.__init__(self)
        self.available = available
        if iconTexture is not None:
            icon = FImage(iconTexture)
            if not available:
                icon.setColour(Color(255, 255, 255, _SHOP_DISABLED_ALPHA))
            self.addChild(icon)
        text = FPlainText(UI.DefaultFont, textValue, _SHOP_TEXT_SIZE)
        text.setLineAlignment(Text.LineAlignment.Right)
        if not available:
            text.setColour(_SHOP_DISABLED_TEXT_COLOUR)
        text.setPosition(
            Vector2f(
                width - _SHOP_TEXT_RIGHT_PAD,
                _SHOP_TEXT_Y_OFFSET,
            )
        )
        self.addChild(text)

    def getChildren(self):
        return []

    def onTick(self, deltaTime: float) -> None:
        r"""\brief Update the cell and render to its internal texture.

        - \param deltaTime Elapsed time in seconds.
        """
        self._buildRenderQueue()
        self.render()


def _loadItemIcon(iconPath: str) -> Optional[Texture]:
    if not iconPath:
        return None
    try:
        if "/" in iconPath or "\\" in iconPath:
            parts = iconPath.replace("\\", "/").split("/")
            if len(parts) >= 2:
                return Manager.loadTexture("/".join(parts[:-1]), parts[-1])
        return Manager.loadTexture("Characters/items", iconPath)
    except Exception:
        return None


class WindowShopCommand(WindowCommand):
    r"""\brief Horizontal buy/sell command bar for the shop."""

    def __init__(
        self,
        rect: Tuple[Pair[int], Pair[int]],
        owner: "WindowShop",
    ) -> None:
        r"""\brief Construct the shop command bar.

        - \param rect The command window rectangle.
        - \param owner The shop coordinator.
        """
        commands = {
            "Buy": {
                "text": LOC("SHOP_BUY"),
                "callback": lambda obj, kwargs: owner.confirmCommand(),
            },
            "Sell": {
                "text": LOC("SHOP_SELL"),
                "callback": lambda obj, kwargs: owner.confirmCommand(),
            },
        }
        super().__init__(rect, commands, rectHeight=32, columns=2)
        self._owner = owner
        self._lastIndex = self.index

    def onTick(self, deltaTime: float) -> None:
        super().onTick(deltaTime)
        if self.index != self._lastIndex:
            self._lastIndex = self.index
            self._owner.setMode(SHOP_MODE_SELL if self.index == 1 else SHOP_MODE_BUY)

    def onKeyDown(self, kwargs: Dict[str, Any]) -> None:
        if Input.isActionTriggered(Input.getCancelKeys(), handled=False):
            self._owner.closeByCancel()
            Input.isActionTriggered(Input.getCancelKeys(), handled=True)
            return
        super().onKeyDown(kwargs)

    def onMouseButtonDown(self, kwargs: Dict[str, Any]) -> bool:
        if kwargs["button"] == Input.Mouse.Button.Right:
            self._owner.closeByCancel()
            return True
        return False


class WindowShopItem(WindowSelectable):
    r"""\brief Two-column shop item list."""

    def __init__(
        self,
        rect: Tuple[Pair[int], Pair[int]],
        owner: "WindowShop",
    ) -> None:
        r"""\brief Construct the shop item list.

        - \param rect The item window rectangle.
        - \param owner The shop coordinator.
        """
        super().__init__(rect, None, None, _SHOP_ITEM_ROW_HEIGHT)
        self._owner = owner
        self._itemIDs: List[str] = []
        self._cellAvailable: List[bool] = []

    def refreshItems(self, itemIDs: List[str], availableMap: Dict[str, bool], valueMap: Dict[str, int]) -> None:
        r"""\brief Rebuild the displayed shop item list.

        - \param itemIDs Ordered item IDs to display.
        - \param availableMap Item availability by ID.
        - \param valueMap Right-side numeric value by ID.
        """
        previousIndex = self.index
        previousItemID = self.getCurrentItemID()
        self._itemIDs = itemIDs
        self._cellAvailable = []
        listView = ListView(self.content.getNoTranslationRect(), _SHOP_ITEM_ROW_HEIGHT, True, _SHOP_ITEM_COLUMNS)
        self.setListView(listView)
        cellWidth = self._getRectWidth()
        itemData = Data.getAllGeneralItemData()
        for itemID in itemIDs:
            member = itemData.get(itemID, {})
            available = availableMap.get(itemID, True)
            self._cellAvailable.append(available)
            cell = _ShopCell(cellWidth, _loadItemIcon(member.get("icon", "")), str(valueMap.get(itemID, 0)), available)
            cell.addConfirmCallback(lambda obj, kwargs: self._owner.confirmItem())
            listView.addChild(cell)
        if len(itemIDs) == 0:
            self.index = None
        elif previousItemID in itemIDs:
            self.index = itemIDs.index(previousItemID)
        elif previousIndex is not None:
            self.index = min(previousIndex, len(itemIDs) - 1)
        else:
            self.index = 0
        if self._rect.getParent() is not None:
            self.content.removeChild(self._rect)

    def getCurrentItemID(self) -> Optional[str]:
        if self.index is None or self.index >= len(self._itemIDs):
            return None
        return self._itemIDs[self.index]

    def isCurrentAvailable(self) -> bool:
        if self.index is None or self.index >= len(self._cellAvailable):
            return False
        return self._cellAvailable[self.index]

    def onKeyDown(self, kwargs: Dict[str, Any]) -> None:
        if Input.isActionTriggered(Input.getCancelKeys(), handled=False):
            self._owner.cancelItemSelection()
            Input.isActionTriggered(Input.getCancelKeys(), handled=True)
            return
        super().onKeyDown(kwargs)

    def onMouseButtonDown(self, kwargs: Dict[str, Any]) -> bool:
        if kwargs["button"] == Input.Mouse.Button.Right:
            self._owner.cancelItemSelection()
            return True
        return False


class WindowShop:
    r"""\brief Integrated shop UI with optional command bar and item list."""

    def __init__(
        self,
        player,
        commandRect: Tuple[Pair[int], Pair[int]] = _DEFAULT_COMMAND_RECT,
        itemRect: Tuple[Pair[int], Pair[int]] = _DEFAULT_ITEM_RECT,
        onClose: Optional[Callable[[], None]] = None,
    ) -> None:
        r"""\brief Construct the shop coordinator.

        - \param player The player whose inventory and GOLD are modified.
        - \param commandRect Rectangle for the buy/sell command bar.
        - \param itemRect Rectangle for the item list window.
        - \param onClose Callback invoked after the shop closes.
        """
        self._player = player
        self._onCloseCallback = onClose
        self._commandWindow = WindowShopCommand(commandRect, self)
        self._itemWindow = WindowShopItem(itemRect, self)
        self._itemTopLeft = Vector2f(float(itemRect[0][0]), float(itemRect[0][1]))
        self._buyItemIDs: List[str] = []
        self._canSell = True
        self._mode = SHOP_MODE_BUY
        self._closed = True
        self.close()

    def getCommandWindow(self) -> WindowShopCommand:
        r"""\brief Get the shop command window.

        - \return The command window.
        """
        return self._commandWindow

    def getItemWindow(self) -> WindowShopItem:
        r"""\brief Get the shop item window.

        - \return The item window.
        """
        return self._itemWindow

    def setPlayer(self, player) -> None:
        r"""\brief Rebind the player instance used by the shop.

        - \param player The new player instance.
        """
        self._player = player

    def getVisible(self) -> bool:
        r"""\brief Return whether the shop UI is visible.

        - \return True when the item window is visible.
        """
        return self._itemWindow.getVisible()

    def isClosed(self) -> bool:
        r"""\brief Return whether the latest shop session has closed.

        - \return True after close.
        """
        return self._closed

    def open(self, buyItemIDs: List[str], canSell: bool) -> None:
        r"""\brief Open the shop with the provided buy list and sell flag.

        - \param buyItemIDs Item IDs available for purchase.
        - \param canSell Whether the sell command is available.
        """
        self._buyItemIDs = self._normalizeBuyItems(buyItemIDs)
        self._canSell = bool(canSell)
        self._mode = SHOP_MODE_BUY
        self._closed = False
        self._commandWindow.index = 0
        self._commandWindow._lastIndex = 0
        self._refreshItems()
        if self._canSell:
            self._commandWindow.setVisible(True)
            self._commandWindow.setActive(True)
            self._itemWindow.setPosition(self._itemTopLeft)
            self._itemWindow.setActive(False)
            self._commandWindow.requestKeyboardFocus()
        else:
            gameSize = GlobalSystem.getGameSize()
            topLeft = Vector2f(float((gameSize.x - _SHOP_ITEM_SIZE) / 2), float((gameSize.y - _SHOP_ITEM_SIZE) / 2))
            self._itemWindow.setPosition(topLeft)
            self._commandWindow.setVisible(False)
            self._commandWindow.setActive(False)
            self._itemWindow.setActive(True)
            self._itemWindow.requestKeyboardFocus()
        self._itemWindow.setVisible(True)

    def close(self) -> None:
        r"""\brief Close and deactivate both shop windows."""
        self._commandWindow.setVisible(False)
        self._commandWindow.setActive(False)
        self._itemWindow.setVisible(False)
        self._itemWindow.setActive(False)
        self._closed = True

    def closeByCancel(self) -> None:
        r"""\brief Close the whole shop via cancel input."""
        Manager.playSE(GameSystem.getCancelSE())
        self._closeAndNotify()

    def setMode(self, mode: str) -> None:
        r"""\brief Switch buy/sell mode and refresh the item list.

        - \param mode The shop mode.
        """
        if mode not in (SHOP_MODE_BUY, SHOP_MODE_SELL):
            return
        if not self._canSell and mode == SHOP_MODE_SELL:
            return
        self._mode = mode
        self._refreshItems()

    def confirmCommand(self) -> None:
        r"""\brief Confirm buy/sell mode and move focus to the item list."""
        self.setMode(SHOP_MODE_SELL if self._commandWindow.index == 1 else SHOP_MODE_BUY)
        Manager.playSE(GameSystem.getDecisionSE())
        self._commandWindow.setActive(False)
        self._itemWindow.setActive(True)
        self._itemWindow.requestKeyboardFocus()

    def cancelItemSelection(self) -> None:
        r"""\brief Cancel item selection, returning to command bar or closing shop."""
        Manager.playSE(GameSystem.getCancelSE())
        if not self._canSell:
            self._closeAndNotify()
            return
        self._itemWindow.setActive(False)
        self._commandWindow.setActive(True)
        self._commandWindow.requestKeyboardFocus()

    def confirmItem(self) -> None:
        r"""\brief Confirm the selected item and perform buy/sell."""
        itemID = self._itemWindow.getCurrentItemID()
        if itemID is None:
            Manager.playSE(GameSystem.getBuzzerSE())
            return
        if self._mode == SHOP_MODE_BUY:
            self._buyItem(itemID)
        else:
            self._sellItem(itemID)

    def _normalizeBuyItems(self, buyItemIDs: List[str]) -> List[str]:
        itemData = Data.getAllGeneralItemData()
        result: List[str] = []
        for itemID in buyItemIDs:
            itemKey = str(itemID)
            if itemKey in itemData and itemKey not in result:
                result.append(itemKey)
        return result

    def _refreshItems(self) -> None:
        if self._mode == SHOP_MODE_BUY:
            itemIDs = self._buyItemIDs
            availableMap = {itemID: self._canBuy(itemID) for itemID in itemIDs}
            valueMap = {itemID: self._getItemPrice(itemID) for itemID in itemIDs}
        else:
            itemIDs = self._getSellableItems()
            availableMap = {itemID: True for itemID in itemIDs}
            valueMap = {itemID: self._player.getItemCount(itemID) for itemID in itemIDs}
        self._itemWindow.refreshItems(itemIDs, availableMap, valueMap)

    def _getSellableItems(self) -> List[str]:
        itemData = Data.getAllGeneralItemData()
        playerItems = self._player._items if hasattr(self._player, "_items") else {}
        result: List[str] = []
        for itemID in itemData.keys():
            if itemID in playerItems and playerItems[itemID] > 0 and self._getItemPrice(itemID) > 0:
                result.append(itemID)
        return result

    def _getItemPrice(self, itemID: str) -> int:
        itemInfo = Data.getGeneralItemData(itemID)
        return int(itemInfo.get("price", 0))

    def _canBuy(self, itemID: str) -> bool:
        return self._player.infoComp.GOLD >= self._getItemPrice(itemID)

    def _buyItem(self, itemID: str) -> None:
        price = self._getItemPrice(itemID)
        if not self._itemWindow.isCurrentAvailable() or self._player.infoComp.GOLD < price:
            Manager.playSE(GameSystem.getBuzzerSE())
            self._refreshItems()
            return
        self._player.infoComp.GOLD -= price
        self._player.addItem(itemID, 1)
        Manager.playSE(GameSystem.getShopSE())
        self._refreshItems()

    def _sellItem(self, itemID: str) -> None:
        price = self._getItemPrice(itemID)
        if price <= 0 or not self._player.removeItem(itemID, 1):
            Manager.playSE(GameSystem.getBuzzerSE())
            self._refreshItems()
            return
        self._player.infoComp.GOLD += price
        Manager.playSE(GameSystem.getShopSE())
        self._refreshItems()

    def _closeAndNotify(self) -> None:
        self.close()
        if self._onCloseCallback is not None:
            self._onCloseCallback()

File: Sample/Scripts/Source/Windows/WindowShop.lua
local Engine = require("Engine")
local GlobalFunctions = require("GlobalFunctions")
local Data = require("Source.Data")
local GameSystem = require("Source.System")
local WindowShopUI = require("Source.UI.WindowShop")
local UiLayout = require("Source.UI.UiLayout")
local WindowShopDetail = require("Source.Windows.WindowShopDetail")
local WindowShopItem = require("Source.Windows.WindowShopItem")
local WindowShopTabs = require("Source.Windows.WindowShopTabs")

local ManagerFunctions = GlobalFunctions.Manager
local Canvas = Engine.Canvas

local _SHOP_TAB_HEIGHT = 64
local _SHOP_ITEM_HEIGHT = 256
local _SHOP_DETAIL_HEIGHT = 96
local _SHOP_WIDTH = 352

---@class Source.Windows.WindowShop
local WindowShop = {}

WindowShop.SHOP_MODE_BUY = "buy"
WindowShop.SHOP_MODE_SELL = "sell"

function WindowShop.GetDefaultRects()
    local totalHeight = _SHOP_TAB_HEIGHT + _SHOP_ITEM_HEIGHT + _SHOP_DETAIL_HEIGHT
    local bounds = UiLayout.GetCenteredRect(_SHOP_WIDTH, totalHeight)
    return Engine.ToIntRect(bounds.position.x, bounds.position.y, _SHOP_WIDTH, _SHOP_TAB_HEIGHT),
        Engine.ToIntRect(bounds.position.x, bounds.position.y + _SHOP_TAB_HEIGHT, _SHOP_WIDTH, _SHOP_ITEM_HEIGHT),
        Engine.ToIntRect(
            bounds.position.x, bounds.position.y + _SHOP_TAB_HEIGHT + _SHOP_ITEM_HEIGHT, _SHOP_WIDTH,
            _SHOP_DETAIL_HEIGHT
        )
end

function WindowShop:init(player, tabRect, itemRect, detailRect, onClose)
    if tabRect == nil or itemRect == nil or detailRect == nil then
        local defaultTabRect, defaultItemRect, defaultDetailRect = WindowShop.GetDefaultRects()
        tabRect = tabRect or defaultTabRect
        itemRect = itemRect or defaultItemRect
        detailRect = detailRect or defaultDetailRect
    end
    local totalHeight = _SHOP_TAB_HEIGHT + _SHOP_ITEM_HEIGHT + _SHOP_DETAIL_HEIGHT
    super(WindowShop, self).init(Engine.ToIntRect(tabRect.position.x, tabRect.position.y, _SHOP_WIDTH, totalHeight))
    self._player = player
    self._onCloseCallback = onClose
    self._ui = WindowShopUI.new(self)
    self._ui:attach()
    self._transition = self._ui:createTransition(self)
    self._tabWindow = WindowShopTabs.new(
        Engine.ToIntRect(0, 0, _SHOP_WIDTH, _SHOP_TAB_HEIGHT), self, self._ui:getTabsAsset()
    )
    self._itemWindow = WindowShopItem.new(
        Engine.ToIntRect(0, _SHOP_TAB_HEIGHT, _SHOP_WIDTH, _SHOP_ITEM_HEIGHT), self, self._ui:getItemAsset()
    )
    self._detailWindow = WindowShopDetail.new(
        Engine.ToIntRect(0, _SHOP_TAB_HEIGHT + _SHOP_ITEM_HEIGHT, _SHOP_WIDTH, _SHOP_DETAIL_HEIGHT),
        self._ui:getDetailAsset()
    )
    self:addChild(self._tabWindow)
    self:addChild(self._itemWindow)
    self:addChild(self._detailWindow)
    self._topLeft = sf.Vector2f.new(tabRect.position.x, tabRect.position.y)
    self._tabTopLeft = sf.Vector2f.new(0.0, 0.0)
    self._itemTopLeft = sf.Vector2f.new(0.0, _SHOP_TAB_HEIGHT)
    self._detailTopLeft = sf.Vector2f.new(0.0, _SHOP_TAB_HEIGHT + _SHOP_ITEM_HEIGHT)
    self._buyItemIDs = {}
    self._canSell = true
    self._mode = self.SHOP_MODE_BUY
    self._closed = true
    self._tabWindow:setVisible(false)
    self._tabWindow:setActive(false)
    self._itemWindow:setVisible(false)
    self._itemWindow:setActive(false)
    self._detailWindow:setVisible(false)
    self._detailWindow:setActive(false)
    self._transition:hideImmediate()
end

function WindowShop:getTabWindow()
    return self._tabWindow
end

function WindowShop:getItemWindow()
    return self._itemWindow
end

function WindowShop:getDetailWindow()
    return self._detailWindow
end

function WindowShop:setPlayer(player)
    self._player = player
end

function WindowShop:getVisible()
    return self._transition:isBlocking()
end

function WindowShop:isClosed()
    return self._closed
end

function WindowShop:open(buyItemIDs, canSell)
    self._buyItemIDs = WindowShop.NormalizeBuyItems(buyItemIDs)
    self._canSell = bool(canSell)
    self._mode = self.SHOP_MODE_BUY
    self._closed = false
    self._tabWindow:getTabView():setSelectedIndex(0)
    self:_refreshItems()
    self._itemWindow:resetSelection()
    self:_refreshDetail()
    if self._canSell then
        self:setPosition(self._topLeft)
        self._tabWindow:setPosition(self._tabTopLeft)
        self._itemWindow:setPosition(self._itemTopLeft)
        self._detailWindow:setPosition(self._detailTopLeft)
        self._tabWindow:setVisible(true)
        self._tabWindow:setActive(true)
    else
        local bounds = UiLayout.GetCenteredRect(_SHOP_WIDTH, _SHOP_ITEM_HEIGHT + _SHOP_DETAIL_HEIGHT)
        self:setPosition(sf.Vector2f.new(bounds.position.x, bounds.position.y))
        self._itemWindow:setPosition(sf.Vector2f.new(0.0, 0.0))
        self._detailWindow:setPosition(sf.Vector2f.new(0.0, _SHOP_ITEM_HEIGHT))
        self._tabWindow:setVisible(false)
        self._tabWindow:setActive(false)
    end
    self._itemWindow:setVisible(true)
    self._itemWindow:setActive(false)
    self._detailWindow:setVisible(true)
    self._detailWindow:setActive(false)
    self._transition:show("FadeIn", function ()
        self:setActive(true)
        if self._canSell then
            self._tabWindow:setActive(true)
        end
        self._itemWindow:setActive(true)
        self._itemWindow:requestKeyboardFocusAtCursor()
    end)
end

function WindowShop:close(onHidden)
    self._tabWindow:setActive(false)
    self._itemWindow:setActive(false)
    self._detailWindow:setActive(false)
    self:setActive(false)
    self._transition:hide("FadeOut", function ()
        self._tabWindow:setVisible(false)
        self._itemWindow:setVisible(false)
        self._detailWindow:setVisible(false)
        self._closed = true
        if onHidden ~= nil then
            onHidden()
        end
    end)
end

function WindowShop:closeByCancel()
    ManagerFunctions.playSE(GameSystem.GetCancelSE())
    self:_closeAndNotify()
end

function WindowShop:handleTabNavigationInput()
    if not self._tabWindow:getVisible() then
        return false
    end
    return self._tabWindow:handleNavigationInput()
end

---@param index integer
function WindowShop:onTabSelected(index)
    self:setMode(index == 1 and self.SHOP_MODE_SELL or self.SHOP_MODE_BUY)
end

function WindowShop:setMode(mode)
    if mode ~= self.SHOP_MODE_BUY and mode ~= self.SHOP_MODE_SELL then
        return
    end
    if not self._canSell and mode == self.SHOP_MODE_SELL then
        return
    end
    if self._mode == mode then
        return
    end
    self._mode = mode
    self:_refreshItems()
    self._itemWindow:resetSelection()
    self:_refreshDetail()
    self._itemWindow:requestKeyboardFocus()
end

function WindowShop:notifyItemIndexMaybeChanged()
    self:_refreshDetail()
end

function WindowShop:refreshLocale()
    self._tabWindow:refresh()
    self._detailWindow:refresh()
end

function WindowShop:confirmItem()
    local itemID = self._itemWindow:getCurrentItemID()
    if itemID == nil then
        ManagerFunctions.playSE(GameSystem.GetBuzzerSE())
        return
    end
    if self._mode == self.SHOP_MODE_BUY then
        self:_buyItem(itemID)
    else
        self:_sellItem(itemID)
    end
end

---@param buyItemIDs table
---@return table
function WindowShop.NormalizeBuyItems(buyItemIDs)
    local itemData = Data.GetAllGeneralItemData()
    local result = {}
    local included = {}
    for _, itemID in ipairs(buyItemIDs) do
        local itemKey = tostring(itemID)
        if itemData[itemKey] ~= nil and not included[itemKey] then
            result[#result + 1] = itemKey
            included[itemKey] = true
        end
    end
    return result
end

function WindowShop:_refreshItems()
    local itemIDs = nil
    local availableMap = {}
    local valueMap = {}
    local showValues = self._mode == self.SHOP_MODE_SELL
    if self._mode == self.SHOP_MODE_BUY then
        itemIDs = self._buyItemIDs
        for _, itemID in ipairs(itemIDs) do
            availableMap[itemID] = self:_canBuy(itemID)
        end
    else
        itemIDs = self:_getSellableItems()
        for _, itemID in ipairs(itemIDs) do
            availableMap[itemID] = true
            valueMap[itemID] = self._player:getItemCount(itemID)
        end
    end
    self._itemWindow:refreshItems(itemIDs, availableMap, valueMap, showValues)
    self:_refreshDetail()
end

function WindowShop:_refreshDetail()
    local itemID = self._itemWindow:getCurrentItemID()
    if itemID == nil then
        self._detailWindow:setItem(nil, nil)
        return
    end
    local price = self._mode == self.SHOP_MODE_BUY and WindowShop.GetItemPrice(itemID)
        or WindowShop.GetSellPrice(itemID)
    self._detailWindow:setItem(Data.GetGeneralItemData(itemID), price)
end

---@return table
function WindowShop:_getSellableItems()
    local itemData = Data.GetAllGeneralItemData()
    local playerItems = self._player._items or {}
    local result = {}
    for _, itemID in ipairs(table.orderedStringKeys(itemData)) do
        if (playerItems[itemID] or 0) > 0 and WindowShop.GetSellPrice(itemID) > 0 then
            result[#result + 1] = itemID
        end
    end
    return result
end

---@param itemID string
---@return integer
function WindowShop.GetItemPrice(itemID)
    local itemInfo = Data.GetGeneralItemData(itemID)
    return itemInfo.price
end

---@param itemID string
---@return integer
function WindowShop.GetSellPrice(itemID)
    return math.floor(WindowShop.GetItemPrice(itemID) / 2)
end

---@param itemID string
---@return boolean
function WindowShop:_canBuy(itemID)
    return self._player.attributes.GOLD >= WindowShop.GetItemPrice(itemID)
end

---@param itemID string
function WindowShop:_buyItem(itemID)
    local price = WindowShop.GetItemPrice(itemID)
    if not self._itemWindow:isCurrentAvailable() or self._player.attributes.GOLD < price then
        ManagerFunctions.playSE(GameSystem.GetBuzzerSE())
        self:_refreshItems()
        return
    end
    local abilitySystem = self._player:getAbilitySystemComponent()
    abilitySystem:setNumericAttributeBase("GOLD", abilitySystem:getNumericAttributeBase("GOLD") - price)
    self._player:addItem(itemID, 1)
    ManagerFunctions.playSE(GameSystem.GetShopSE())
    self:_refreshItems()
end

---@param itemID string
function WindowShop:_sellItem(itemID)
    local price = WindowShop.GetSellPrice(itemID)
    if price <= 0 or not self._player:removeItem(itemID, 1) then
        ManagerFunctions.playSE(GameSystem.GetBuzzerSE())
        self:_refreshItems()
        return
    end
    local abilitySystem = self._player:getAbilitySystemComponent()
    abilitySystem:setNumericAttributeBase("GOLD", abilitySystem:getNumericAttributeBase("GOLD") + price)
    ManagerFunctions.playSE(GameSystem.GetShopSE())
    self:_refreshItems()
end

function WindowShop:_closeAndNotify()
    self:close(function ()
        if self._onCloseCallback ~= nil then
            self._onCloseCallback()
        end
    end)
end

function WindowShop:dispose()
    self._transition:hideImmediate()
    self._tabWindow:setVisible(false)
    self._itemWindow:setVisible(false)
    self._detailWindow:setVisible(false)
    self._tabWindow:dispose()
    self._itemWindow:dispose()
    self._detailWindow:dispose()
    self._ui:dispose()
    self._player = nil
    self._onCloseCallback = nil
end

return class(WindowShop, Canvas)
